pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -582,6 +595,83 @@ func (gce *GCECloud) IsLegacyNetwork() bool {
 	return gce.isLegacyNetwork
 }
 
+func (gce *GCECloud) SetInformers(informerFactory informers.SharedInformerFactory) {
+	glog.Infof("Setting up informers for GCECloud")
+	nodeInformer := informerFactory.Core().V1().Nodes()
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			node := obj.(*v1.Node)
+			gce.insertNodeZones(node)
+		},
+		UpdateFunc: func(prev, obj interface{}) {
+			prevNode := prev.(*v1.Node)
+			newNode := obj.(*v1.Node)
+			gce.updateNodeZones(prevNode, newNode)
+		},
+		DeleteFunc: func(obj interface{}) {
+			node, isNode := obj.(*v1.Node)
+			// We can get DeletedFinalStateUnknown instead of *v1.Node here and we need to handle that correctly.
+			if !isNode {
+				deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					glog.Errorf("Received unexpected object: %v", obj)
+					return
+				}
+				node, ok = deletedState.Obj.(*v1.Node)
+				if !ok {
+					glog.Errorf("DeletedFinalStateUnknown contained non-Node object: %v", deletedState.Obj)
+					return
+				}
+			}
+			gce.deleteNodeZones(node)
+		},
+	})
+	gce.nodeInformerSynced = nodeInformer.Informer().HasSynced
+}
+
+func (gce *GCECloud) updateNodeZones(prevNode, newNode *v1.Node) {
+	gce.nodeInformerLock.Lock()
+	defer gce.nodeInformerLock.Unlock()
+	prevZone, ok := prevNode.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+	if ok {
+		gce.nodeZones[prevZone].Delete(prevNode.ObjectMeta.Name)
+		if gce.nodeZones[prevZone].Len() == 0 {
+			gce.nodeZones[prevZone] = nil
+		}
+	}
+	newZone, ok := newNode.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+	if ok {
+		if gce.nodeZones[newZone] == nil {
+			gce.nodeZones[newZone] = sets.NewString()
+		}
+		gce.nodeZones[newZone].Insert(newNode.ObjectMeta.Name)
+	}
+}
+
+func (gce *GCECloud) insertNodeZones(node *v1.Node) {
```

Comment:
Instead of having all:
- insertNodeZones
- updateNodeZones
- deleteNodeZones
[which all look pretty much the same]

it would be enough to have "updateNodeZones" (as above), but allow for passing "nil" as both prevNode and newNode

[Then prevNode=nil means insert, newNode=nil means delete]